flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/memory/MemoryTierSubpartitionProducerAgent.java

```diff
@@ -0,0 +1,101 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.partition.hybrid.tiered.tier.memory;
+
+import org.apache.flink.core.memory.MemorySegment;
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.netty.NettyConnectionWriter;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.netty.NettyPayload;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.netty.TieredStorageNettyService;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.netty.TieredStorageNettyServiceImpl;
+
+import java.nio.ByteBuffer;
+
+import static org.apache.flink.runtime.io.network.buffer.Buffer.DataType.END_OF_SEGMENT;
+
+/** The subpartition producer agent for the memory tier. */
+class MemoryTierSubpartitionProducerAgent {
+
+    private final int subpartitionId;
+
+    private final TieredStorageNettyService nettyService;
+
+    private NettyConnectionWriter nettyConnectionWriter;
+
+    private int finishedBufferIndex;
+
+    MemoryTierSubpartitionProducerAgent(
+            int subpartitionId, TieredStorageNettyService nettyService) {
+        this.subpartitionId = subpartitionId;
+        this.nettyService = nettyService;
+    }
+
+    // ------------------------------------------------------------------------
+    //  Called by MemoryTierProducerAgent
+    // ------------------------------------------------------------------------
+
+    void registerNettyService(NettyConnectionWriter nettyConnectionWriter) {
+        this.nettyConnectionWriter = nettyConnectionWriter;
+    }
+
+    void addFinishedBuffer(Buffer buffer) {
+        NettyPayload toAddBuffer =
+                NettyPayload.newBuffer(buffer, finishedBufferIndex, subpartitionId);
+        addFinishedBuffer(toAddBuffer);
+    }
+
+    void addEndOfSegmentEvent(ByteBuffer event) {
+        MemorySegment data = MemorySegmentFactory.wrap(event.array());
+        Buffer buffer =
+                new NetworkBuffer(
+                        data, FreeingBufferRecycler.INSTANCE, END_OF_SEGMENT, data.size());
+        NettyPayload nettyPayload =
+                NettyPayload.newBuffer(buffer, finishedBufferIndex, subpartitionId);
+        addFinishedBuffer(nettyPayload);
+    }
```

Comment:
It's weird that both `MemoryTierProducerAgent` and `MemoryTierSubpartitionProducerAgent` need to be aware of the details end-of-segment event. Can we prepare the event outside the subpartition agent?

Reply:
Fixed. Only the `MemoryTierProducerAgent ` need to be aware of the end of segment now.